include/eve/algo/container/detail/soa_storage2.hpp

```diff
@@ -0,0 +1,177 @@
+#pragma once
+
+#include <eve/module/core.hpp>
+#include <eve/algo/views/zip.hpp>
+#include <eve/detail/kumi.hpp>
+#include <eve/detail/raberu.hpp>
+#include <eve/memory/aligned_allocator.hpp>
+#include <eve/concept/simd_allocator.hpp>
+#include <eve/product_type.hpp>
+#include <type_traits>
+#include <memory>
+
+namespace eve::algo::detail
+{
+  template<eve::product_type Type, eve::simd_allocator Allocator>
+  struct soa_storage : Allocator
+  {
+    using is_product_type = void;
+
+              soa_storage() : storage_( nullptr ), size_(0) { build(0); }
+    explicit  soa_storage( std::size_t n) : soa_storage( n, Type{} ) {}
+
+    soa_storage( std::size_t n, Type const& v)  : soa_storage()
+    {
+      storage_ = (byte_t*)Allocator::aligned_alloc(byte_size(n), 64);
+      size_ = n;
+      build(n,v);
+    }
+
+    soa_storage(soa_storage const& src) : soa_storage()
+    {
+      storage_ = (byte_t*)Allocator::aligned_alloc(size_, 64);
+      size_ = src.size_;
+      std::copy(storage_, storage_+size_, src.storage_);
+      reindex(src.size_);
+    }
+
+    soa_storage& operator=(soa_storage const& src)
+    {
+      soa_storage that(src);
+      swap(that);
+      return *this;
+    }
+
+    soa_storage(soa_storage&&)            =default;
+    soa_storage& operator=(soa_storage&&) =default;
+
+    ~soa_storage()
+    {
+      kumi::for_each( [&]<typename T>(T*& s)
+                      {
+                        for(std::size_t i=0;i<size_;++i) static_cast<T*>(s+i)->~T();
+                      }
+                    , indexes_
+                    );
+      Allocator::aligned_dealloc(storage_, size_);
+    }
+
+    EVE_FORCEINLINE auto data_aligned()
+    {
+      auto ptrs = kumi::map([](auto* m) { return as_aligned_pointer(m); }, indexes_);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data_aligned() const
+    {
+      auto ptrs = kumi::map([](auto const* m) { return as_aligned_pointer(m); }, indexes_);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data()       { return eve::unalign(data_aligned()); }
+    EVE_FORCEINLINE auto data() const { return eve::unalign(data_aligned()); }
+
+    void swap(soa_storage& other) noexcept
+    {
+      std::swap(storage_,other.storage_);
+      std::swap(indexes_,other.indexes_);
+      std::swap(size_,other.size_);
+    }
+
+    friend void swap(soa_storage& a, soa_storage& b ) noexcept
+    {
+      a.swap(b);
+    }
+
+    void build(std::size_t n, Type v = {})
+    {
+      auto sub_span = storage_;
+
+      kumi::for_each( [&]<typename T>(T*& s, auto m)
+                      {
+                        s = reinterpret_cast<T*>(sub_span);
+                        for(std::size_t i=0;i<n;++i) new (s+i) T(m);
+                        sub_span += realign<T>(n);
+                      }
+                    , indexes_, v
+                    );
```

Comment:
We really. need tests for not flat structs. 

I do not like that this isn't done with simd. We should be able to use eve::algo::fill and make it legal.